rpc/core/bridge_commitments.go

```diff
@@ -0,0 +1,232 @@
+package core
+
+import (
+	"encoding/hex"
+	"fmt"
+	"strconv"
+
+	"github.com/cometbft/cometbft/crypto/merkle"
+	ctypes "github.com/cometbft/cometbft/rpc/core/types"
+	rpctypes "github.com/cometbft/cometbft/rpc/jsonrpc/types"
+	"github.com/cometbft/cometbft/types"
+)
+
+const (
+	// BridgeCommitmentBlocksLimit is the limit to the number of blocks we can generate a bridge commitment for.
+	// This limits the ZK prover time required to compute.
+	BridgeCommitmentBlocksLimit = 1000
+)
+
+// BridgeCommitment collects the transactions results roots over a provided ordered range of blocks,
+// and then creates a new merkle root. The range is end exclusive.
+func (env *Environment) BridgeCommitment(_ *rpctypes.Context,
+	start, end uint64,
+) (*ctypes.ResultBridgeCommitment, error) {
+	err := env.validateBridgeCommitmentRange(start, end)
+	if err != nil {
+		return nil, err
+	}
+
+	// Fetch data.
+	leaves, err := env.fetchBridgeCommitmentLeaves(start, end)
+	if err != nil {
+		return nil, err
+	}
+	// Encode data to match solidity `abi.encode`.
+	encodedLeaves, err := env.encodeBridgeCommitment(leaves)
+	if err != nil {
+		return nil, err
+	}
+	root := merkle.HashFromByteSlices(encodedLeaves)
+
+	return &ctypes.ResultBridgeCommitment{
+		BridgeCommitment: root,
+	}, nil
+}
+
+// BridgeCommitmentInclusionProof creates two inclusion proofs to verify that a transaction in included
+// in a BridgeCommitment. Users can also verify any data in the transactions responses for data
+// availability. Users need to provide the height and the transaction index that the inclusion proof is
+// for. They also need to provide the indexes of the start and end blocks for which the BridgeCommitment
+// merkle root is constructed from. The range for BridgeCommitment is end exclusive.
+func (env *Environment) BridgeCommitmentInclusionProof(
+	_ *rpctypes.Context,
+	height, txIndex int64,
+	start, end uint64,
+) (*ctypes.ResultBridgeCommitmentInclusionProof, error) {
+	err := env.validateBridgeCommitmentInclusionProofRequest(uint64(height), start, end)
+	if err != nil {
+		return nil, err
+	}
+
+	// Fetch data.
+	leaves, err := env.fetchBridgeCommitmentLeaves(start, end)
+	if err != nil {
+		return nil, err
+	}
+	// Encode data to match solidity `abi.encode`.
+	encodedLeaves, err := env.encodeBridgeCommitment(leaves)
+	if err != nil {
+		return nil, err
+	}
+	// Get proofs of the BridgeCommitment leaves.
+	_, proofs := merkle.ProofsFromByteSlices(encodedLeaves)
+	bcProof := proofs[height-int64(leaves[0].Height)]
+
+	// Load the transactions that composed the LastResultsHash at height.
+	finalizeBlockResponse, err := env.StateStore.LoadFinalizeBlockResponse(height - 1)
```

Comment:
So if a transaction was processed at `h` but produced a result at `h+1`, the bridge commitment inclusion call should be `BridgeCommitmentInclusionProof(_, h+1, _, _, _)`? Ignore the underscores.

Reply:
If so, it feels a bit odd because there's cases where we're getting the proof for a transaction that falls outside of a block range:

`BridgeCommitmentInclusionProof(_, h+1, 0, h+1, h+101)`

This will give us a proof for a transaction `0` at `h`.

Reply:
I think we can just clarify in the function documentation.

From:

`...to verify that a transaction in included in a BridgeCommitment`

To:

`...to verify that a transaction response is included in a BridgeCommitment`